src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -422,28 +506,59 @@ private Type EmitIdentifierListNode(AssociativeNode node)
 
         private Type EmitExprListNode(AssociativeNode node)
         {
-            if (compilePass == CompilePass.GlobalFuncSig) return null;
+            if (compilePass == CompilePass.MethodLookup) return null;
 
             if (!(node is ExprListNode eln))
             {
                 throw new ArgumentException("AST node must be an Expression List.");
             }
+            var arrayTypes = GetTypeStatisticsForArray(eln);
+            var ot = GetOverallTypeForArray(arrayTypes);
+
             var elements = eln.Exprs;
+            
             EmitOpCode(OpCodes.Ldc_I4, elements.Count);
-            EmitOpCode(OpCodes.Newarr, typeof(double));
+            EmitOpCode(OpCodes.Newarr, ot);
             int elCount = -1;
             foreach (var el in elements)
             {
                 EmitOpCode(OpCodes.Dup);
                 EmitOpCode(OpCodes.Ldc_I4, ++elCount);
                 var t = DfsTraverse(el);
-                //if (t == typeof(int) || t == typeof(long) || t == typeof(double) || t == typeof(bool) || t == typeof(char))
-                //{
-                //    EmitOpCode(OpCodes.Box, t);
-                //}
-                EmitOpCode(OpCodes.Stelem_R8);
+                if (ot == typeof(object) && (t == typeof(int) || t == typeof(long) || t == typeof(double) || t == typeof(bool) || t == typeof(char)))
```

Comment:
So is the idea here to check for non-array types ?

If that is the case...could we not use the `IsArray` property of the `Type` class ?

https://docs.microsoft.com/en-us/dotnet/api/system.type.isarray?redirectedfrom=MSDN&view=net-6.0#System_Type_IsArray

Reply:
The idea here is to check for primitive (value) types to see if they need to be boxed to convert to reference type.

Reply:
I think there is a `IsValueType` property on Types